guardian/mobile-n10n
# LIVE-4736 add configuration and logging to EC2 sender worker

## What does this change?

We are building a EC2-based sender worker to further improve the performance of the notification delivery pipeline.

After PR #793 and PR  #796, we have a subproject which creates a EC2-based cloudformation stack and deploys a _empty_ `sender-worker` via `riff-raff`.

The PR wires up the configuration setting and logging facility and makes changes so that the sender worker lambda functions can be reused directly in the EC2-based sender.

## Logging

The _slf4j_ logging configuration is defined in `resources/logback.xml`.  Like the registrations API / notificatIons API, it writes logs to a log file on the file system (`sender-worker/sender-worker.log`).  In the EC2 instance, a `aws-kinesis-agent` is used to copy the logs from the log file to the Kinesis stream `mobile-log-aggregate-$(stage)`.  Logs are then shipped to Kibana.

## Configuration

It seems reasonable to expect that the lambda-based sender and EC2-based sender may have different configuration due to the difference in their underlying infrastructure.  A new set of configuration has been created for the EC2-based sender under the SSM parameter store `notifications/$(stage)/ec2workers`, but it has the same structure of parameter keys as lambda-based sender's.

## Local Development

Local deployment is enabled.  It requires a local configuration file to store the whole set of parameters.  A shell script (`scripts/create-local-conf-from-CODE.sh`) is provided to help creating one from the parameters in `CODE`:

1. Execute the shell script by: ```. ./scripts/create-local-conf-from-CODE.sh```
2. Copy the output to `~/.gu/sender-worker.conf`.  Note that it contains secret data such as the Firebase account key and APNS certificates.

The two steps above need to be done once only.

4. Run 'sbt'
5. Run 'project sender-worker` in `sbt` shell
6. Run `run`
 
## Changes to `notificationworkerlambda`

To allow the EC2 sender to use the sender functions defined in `notificationworkerlambda` directly, the following changes have been made to the lambda sender:

1. The `IOSSender` class has been changed to have another constructor which takes `ApnsWorkerConfiguration` config from the argument. 

2. The `AndroidSender` class has been changed to have another constructor which takes `FcmWorkerConfiguration` config and an optional FirebaseApp name from the argument.

3. The `FcmClient.apply` method accepts a new optional argument for the name of FirebaseApp.  Each FirebaseApp instance can be initialised once only, but for EC2 sender, it may instantiate multiple`FirebaseApp` for `Android`, `AndroidBeta` and `AndroidEdition`.  In this case, we need to initialise FirebaseApp with a unique name (unique within this process).

## How to test

The `AndroidSender` and `IOSSender` were instantiated successfully.  The sender attempted to send a test `Notification`  to an invalid token and the FIrebase / APNS returns an expected error of invalid token in both `CODE` and local environment.

Logs from `sender-worker` `CODE` can be found in `Kibana`.  Logs from local deployment can be found in `log/sender-worker.log`.

I will run a test of sending notifications on `CODE` to make sure that it has no impact on the existing lambda sender worker.
